src/features/devices/hooks/target-hostapp.ts

```diff
@@ -0,0 +1,277 @@
+			req: permissions.root,
+		},
+	});
+
+	return Promise.all(
+		_.map(devicesByDeviceType, async (affectedDevices, deviceType) => {
+			const affectedDeviceIds = affectedDevices.map((d) => d.id);
+
+			const [osRelease] = await getOSReleaseResource(
+				api,
+				osVersion,
+				osVariant,
+				deviceType,
+			);
+
+			if (osRelease == null) {
+				return;
+			}
+
+			await rootApi.patch({
+				resource: 'device',
+				options: {
+					$filter: {
+						id: { $in: affectedDeviceIds },
+					},
+				},
+				body: {
+					is_initialized_by__release: osRelease.id,
+				},
+			});
+		}),
+	);
+}
+
+async function getOSReleaseResource(
+	api: sbvrUtils.PinejsClient,
+	osVersion: string,
+	osVariant: string,
+	deviceTypeId: string,
+) {
+	return await api.get({
+		resource: 'release',
+		options: {
+			$select: ['id', 'belongs_to__application'],
+			$filter: {
+				// TODO: better to (eventually) use release version, once that's fully supported
+				// https://www.flowdock.com/app/rulemotion/resin-tech/threads/Ao4qzbDh8Z4Pgq_xF99Ld5fME35
+				status: 'success',
+				belongs_to__application: {
+					$any: {
+						$alias: 'a',
+						$expr: {
+							$and: [
+								{
+									a: {
+										is_for__device_type: deviceTypeId,
+									},
+								},
+								{
+									a: {
+										is_host: true,
+									},
+								},
+							],
+						},
+					},
+				},
+				$and: [
+					{
+						release_tag: {
+							$any: {
+								$alias: 'rt',
+								$expr: {
+									rt: {
+										// TODO: probably better to move all this string munging to a normalization
+										// function. also, maybe we care about devices provisioning with `resinOS` in
+										// the name, but probably we can just treat those as so old it doesn't really
+										// matter.
+										value: osVersion.replace('balenaOS ', ''),
+										tag_key: 'version',
+									},
+								},
+							},
+						},
+					},
+					{
+						release_tag: {
+							$any: {
+								$alias: 'rt',
+								$expr: {
+									rt: {
+										// TODO: probably better to move all this string munging to a normalization
+										// function.
+										value: osVariant === 'prod' ? 'production' : 'development',
+										tag_key: 'variant',
+									},
+								},
+							},
+						},
+					},
+				],
+			},
+		},
+	});
+}
+
+async function checkHostappReleaseUpgrades(
+	api: sbvrUtils.PinejsClient,
+	deviceIds: number[],
+	newHostappReleaseId: number,
+) {
+	if (deviceIds.length === 0) {
+		return;
+	}
+	const nullHostappCount = await api.get({
+		resource: 'device',
+		options: {
+			$count: { $filter: { id: { $in: deviceIds }, os_version: null } },
+		},
+	});
+
+	if (nullHostappCount === deviceIds.length) {
+		return;
+	}
+
+	const newHostappRelease = await api.get({
+		resource: 'release',
+		id: newHostappReleaseId,
+		options: {
+			$expand: {
+				release_tag: { $filter: { tag_key: 'version' }, $select: ['value'] },
+			},
+			$filter: {
+				is_invalidated: false,
```

Comment:
we shouldn't be able to _upgrade_ to an invalid release, but we can provision to one (why it's not an SBVR, but filtered here).

Reply:
do you mean express this as a rule, or something different?

Reply:
Well expressing precisely what we allow would be the first step, with the ideal outcome being that it can be expressed as an SBVR rule (and if we can't then I expect it'll be due to an element missing from our model rather than a fundamental blocker)

Reply:
this is an issue we've discussed in the past generally speaking -- that we don't want to prevent a device from coming online with a known-bad version (i.e. creates are allowed), but we want to prevent other devices from upgrading to them (updates are disallowed). this is true for both user apps as well as infrastructure apps like the supervisor or hostapp.

Reply:
discussion about converting this into a rule here: https://www.flowdock.com/app/rulemotion/resin-tech/threads/Fvz_ycKmosdRhCX61uMLmXlK0o-